dev/tools/gen_localizations.dart

```diff
@@ -107,19 +107,74 @@ String generateString(String s) {
   return output.toString();
 }
 
+/// Simple data class to hold parsed locale. Does not promise validity of any data.
+class Locale {
+  Locale({
+    this.languageCode,
+    this.scriptCode,
+    this.countryCode,
+    this.length,
+    this.origString
+  });
+
+  String languageCode;
+  String scriptCode;
+  String countryCode;
+  int length;          // The number of fields. Ranges from 1-3.
+  String origString;   // Original un-parsed locale string.
+}
+
+/// Simple parser. Expects the locale string to be in the form of 'language_script_COUNTRY'
+/// where the langauge is 2 characters, script is 4 characters with the first uppercase,
+/// and country is 2-3 characters and all uppercase.
+///
+/// 'language_COUNTRY' or 'language_script' are also valid. Missing fields will be null.
+Locale parseLocaleString(String locale) {
+    final List<String> codes = locale.split('_'); // [language, script, country]
+    String scriptCode;
+    String countryCode;
+    if (codes.length == 2) {
+      scriptCode = codes[1].length >= 4 ? codes[1] : null;
+      countryCode = codes[1].length < 4 ? codes[1] : null;
+    } else if (codes.length == 3) {
+      scriptCode = codes[1].length > codes[2].length ? codes[1] : codes[2];
+      countryCode = codes[1].length < codes[2].length ? codes[1] : codes[2];
+    }
+    assert(codes.length == 1 || codes.length == 2 || codes.length == 3);
```

Comment:
This assert should appear right below the `codes` var. Could write the assert like this:
```
assert(codes.isNotEmpty && codes.length < 4);
```
I imagine that we'd want to throw out locale strings "" and "_" and maybe some other somethings-terribly-wrong cases.  We could apply trim to the language etc parameters and have the local Locale object assert that those parameters are either null or isNotEmpty. Except languageCode which should just be isNotEmpty.